substrate/frame/delegated-staking/src/tests.rs

```diff
@@ -0,0 +1,647 @@
+			Error::<T>::NothingToSlash
+		);
+	});
+}
+
+/// Integration tests with pallet-staking.
+mod staking_integration {
+	use super::*;
+	use pallet_staking::RewardDestination;
+	use sp_staking::Stake;
+
+	#[test]
+	fn bond() {
+		ExtBuilder::default().build_and_execute(|| {
+			let agent: AccountId = 99;
+			let reward_acc: AccountId = 100;
+			assert_eq!(Staking::status(&agent), Err(StakingError::<T>::NotStash.into()));
+
+			// set intention to become an agent
+			fund(&agent, 100);
+			assert_ok!(DelegatedStaking::register_agent(
+				RawOrigin::Signed(agent).into(),
+				reward_acc
+			));
+			assert_eq!(DelegatedStaking::stakeable_balance(&agent), 0);
+
+			let mut delegated_balance: Balance = 0;
+
+			// set some delegations
+			for delegator in 200..250 {
+				fund(&delegator, 200);
+				assert_ok!(DelegatedStaking::delegate_to_agent(
+					RawOrigin::Signed(delegator).into(),
+					agent,
+					100
+				));
+				delegated_balance += 100;
+				assert_eq!(
+					Balances::balance_on_hold(&HoldReason::StakingDelegation.into(), &delegator),
+					100
+				);
+				assert_eq!(DelegatedStaking::delegator_balance(&delegator), 100);
+
+				let agent_obj = get_agent(&agent);
+				assert_eq!(agent_obj.ledger.stakeable_balance(), delegated_balance);
+				assert_eq!(agent_obj.available_to_bond(), 0);
+				assert_eq!(agent_obj.bonded_stake(), delegated_balance);
+			}
+
+			assert_eq!(Staking::stake(&agent).unwrap(), Stake { total: 50 * 100, active: 50 * 100 })
+		});
+	}
+
+	#[test]
+	fn withdraw_test() {
+		ExtBuilder::default().build_and_execute(|| {
+			// initial era
+			start_era(1);
+			let agent: AccountId = 200;
+			let reward_acc: AccountId = 201;
+			let delegators: Vec<AccountId> = (301..=350).collect();
+			let total_staked =
+				setup_delegation_stake(agent, reward_acc, delegators.clone(), 10, 10);
+
+			// lets go to a new era
+			start_era(2);
+
+			assert!(eq_stake(agent, total_staked, total_staked));
+			// Withdrawing without unbonding would fail.
+			assert_noop!(
+				DelegatedStaking::release_delegation(RawOrigin::Signed(agent).into(), 301, 50, 0),
+				Error::<T>::NotEnoughFunds
+			);
+
+			// 305 wants to unbond 50 in era 2, withdrawable in era 5.
+			assert_ok!(Staking::unbond(RawOrigin::Signed(agent).into(), 50));
+
+			// 310 wants to unbond 100 in era 3, withdrawable in era 6.
+			start_era(3);
+			assert_ok!(Staking::unbond(RawOrigin::Signed(agent).into(), 100));
+
+			// 320 wants to unbond 200 in era 4, withdrawable in era 7.
+			start_era(4);
+			assert_ok!(Staking::unbond(RawOrigin::Signed(agent).into(), 200));
+
+			// active stake is now reduced..
+			let expected_active = total_staked - (50 + 100 + 200);
+			assert!(eq_stake(agent, total_staked, expected_active));
+
+			// nothing to withdraw at era 4
+			assert_noop!(
+				DelegatedStaking::release_delegation(RawOrigin::Signed(agent).into(), 305, 50, 0),
+				Error::<T>::NotEnoughFunds
+			);
```

Comment:
assert_noop implies this, but okay to be verbose

Reply:
right, will remove the assertion for stake which I check before the call.
The other two checks are probably ok to stay, shows the ongoing setup and wasn't asserted before.